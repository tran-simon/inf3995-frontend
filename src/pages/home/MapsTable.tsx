import MapData from '../../model/MapData';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  IconButton,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableProps,
  TableRow,
} from '@material-ui/core';
import { format, isValid } from 'date-fns';
import { frCA } from 'date-fns/locale';
import DeleteOutlineIcon from '@material-ui/icons/DeleteOutline';
import {
  FirebaseDatabaseMutation,
  FirebaseDatabaseNode,
} from '@react-firebase/database';
import React, { useState } from 'react';
import { ObjectWithKey } from '../../model';
import { useHistory } from 'react-router-dom';

interface MapsTableProps extends TableProps {
  mapsData: ObjectWithKey<MapData>[];
}

const MapsTable = ({ mapsData, ...props }: MapsTableProps) => {
  const [deleteMapKey, setDeleteMapKey] = useState<string | null>(null);
  const history = useHistory();

  return (
    <>
      <Table {...props}>
        <TableHead>
          <TableRow>
            <TableCell>Indice</TableCell>
            <TableCell>Date de dernière modification</TableCell>
            <TableCell>Actions</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {mapsData.map((map, i) => (
            <TableRow
              key={i}
              hover
              onClick={() => {
                history.push(`/explore/${map._key}`);
              }}
            >
              <TableCell>{i}</TableCell>
              <TableCell>
                {isValid(map.date) &&
                  format(map.date, 'PPPP', { locale: frCA })}
              </TableCell>
              <TableCell>
                <IconButton
                  onClick={() => {
                    setDeleteMapKey(map._key);
                  }}
                >
                  <DeleteOutlineIcon />
                </IconButton>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
      <Dialog open={!!deleteMapKey} onClose={() => setDeleteMapKey(null)}>
        {!!deleteMapKey && (
          <FirebaseDatabaseMutation path={`maps/${deleteMapKey}`} type={'set'}>
            {({ runMutation }) => (
              <>
                <DialogTitle>Êtes-vous certain?</DialogTitle>
                <DialogContent>
                  <DialogContentText>
                    Cette action est irréversible
                  </DialogContentText>
                </DialogContent>
                <DialogActions>
                  <Button onClick={() => setDeleteMapKey(null)}>Annuler</Button>
                  <Button
                    onClick={() => {
                      runMutation(null).then(() => {
                        setDeleteMapKey(null);
                      });
                    }}
                  >
                    Confirmer
                  </Button>
                </DialogActions>
              </>
            )}
          </FirebaseDatabaseMutation>
        )}
      </Dialog>
    </>
  );
};

const resValueToMapsData = (resValue: any): ObjectWithKey<MapData>[] => {
  return Object.keys(resValue).map((key) => ({
    _key: key,
    ...resValue[key],
  }));
};

export default () => {
  return (
    <FirebaseDatabaseNode path={'/maps'} orderByChild="date">
      {(res) => {
        const mapsData: ObjectWithKey<MapData>[] =
          (res.isLoading != null &&
            !res.isLoading &&
            res.value &&
            resValueToMapsData(res.value)) ||
          [];

        return <MapsTable mapsData={mapsData} />;
      }}
    </FirebaseDatabaseNode>
  );
};
