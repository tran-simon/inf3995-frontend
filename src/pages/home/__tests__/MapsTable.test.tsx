import { render, waitFor } from '@testing-library/react';
import React from 'react';
import MapsTable from '../MapsTable';

jest.mock('@react-firebase/database', () => ({
  FirebaseDatabaseNode: ({ children }) =>
    children({
      isLoading: false,
      value: {
        id_1: {
          date: 1617917731321,
          name: 'name_1',
        },
        id_2: {
          date: 1617917735555,
          name: 'name_2',
          simulation: true,
        },
      },
    }),
  FirebaseDatabaseMutation: ({ children }) =>
    children({
      runMutation: jest.fn(),
    }),
}));

describe('MapsTable', () => {
  it('can match snapshot', async () => {
    await waitFor(() => {
      expect(render(<MapsTable />).asFragment()).toMatchSnapshot();
    });
  });
});
