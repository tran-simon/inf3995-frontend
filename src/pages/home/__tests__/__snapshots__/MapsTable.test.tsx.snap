// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapsTable can match snapshot 1`] = `
<DocumentFragment>
  <table
    class="MuiTable-root"
  >
    <thead
      class="MuiTableHead-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-head"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Indice
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Nom
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Date de dernière modification
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody
      class="MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-hover"
        style="cursor: pointer;"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          0
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          name_2
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          jeudi 8 avril 2021 à 17:35:35
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
          </button>
        </td>
      </tr>
      <tr
        class="MuiTableRow-root MuiTableRow-hover"
        style="cursor: pointer;"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          1
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          name_1
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          jeudi 8 avril 2021 à 17:35:31
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
